### ISSUE TITLE
change many linked properties to edge relations

### ISSUE THREAD

**Issue by joshStillerman** [joshStillerman/NDM#34]
As in commentOn, need to change tape/reel, etc... to be edges instead of link properties.

To this end I have added:
cutFrom.json

How do we specify in schema that the endpoints must be 
out link -> reel
and
  in linset - tape

?
In the orient studio this is done by saying that the reel has an **in** which is a linkset which must be a cutFrom, and tape has an **out** link which must be a cutFrom.

Not sure this would prevent other things from having cutFrom - needs to be further investigated

**Comment by joshStillerman** [comment]
misunderstood a bit:

the in and out need to be mandatory and properties of the edge not the vertices.  Makes more sense.

```
orientdb> create property cutFrom.out LINK reel
orientdb> create property cutFrom.in LINK tape
orientdb> alter property cutFrom.out MANDATORY=true
orientdb> alter property cutFrom.in MANDATORY=true
orientdb> create index UniqueCutFrom on cutFrom(out,in) unique

```
MANDATORY makes the DB use 'non lightweight edges'
unique index means that each tape is cut from at most 1 reel

**Comment by joshStillerman** [comment]
need relations for
```
  cutFrom (tape, reel)
  shotIn(shot, runDay)

```
should tapeStarData  be an edge or a property ?  Property I think.

Need to remove link properties that correspond to these

**Comment by jcwright77** [comment]
That is the way to enforce link targets. I’ve tested it. What is not clear is how to allow multiple different targets. For example? Will subclasses be also permitted as targets? How about different unrelated targets. I think the answer is this is not supported yet.

I’ll check if this feature request has been made. Could have a fn to make links that would enforce this or maybe a hook for edge creation.

-john

On Mar 6, 2018, at 4:29 PM, Josh Stillerman <notifications@github.com<mailto:notifications@github.com>> wrote:


As in commentOn, need to change tape/reel, etc... to be edges instead of link properties.

To this end I have added:
cutFrom.json

How do we specify in schema that the endpoints must be
out link -> reel
and
in linset - tape

?
In the orient studio this is done by saying that the reel has an in which is a linkset which must be a cutFrom, and tape has an out link which must be a cutFrom.

Not sure this would prevent other things from having cutFrom - needs to be further investigated

**Comment by joshStillerman** [comment]
If the in and out of the edge are of type link it will enforce it.

and each edge connects exactly 2 things so it is not an issue. Just need 
syntax and code in addclass that puts these properties into the edge 
when the class is created.

**Comment by jcwright77** [comment]
I meant multiple different types. Not multiple at once. If you constrain edge to one class that is all it can connect to. Which may be ok. For commentsOn we will probably not want to do this.

-john

On Mar 6, 2018, at 7:20 PM, Josh Stillerman <notifications@github.com<mailto:notifications@github.com>> wrote:

If the in and out of the edge are of type link it will enforce it.

and each edge connects exactly 2 things so it is not an issue. Just need
syntax and code in addclass that puts these properties into the edge
when the class is created.

**Comment by joshStillerman** [comment]
We can constrain it to be an _NDMOBJECT

and for other things we can use super classes and constrain to those I think

**Comment by jcwright77** [comment]
Yes. As I said if it is possible that is how to do It. That may work, but I expect it won’t.  Have you tested?

-john

On Mar 6, 2018, at 7:38 PM, Josh Stillerman <notifications@github.com<mailto:notifications@github.com>> wrote:

We can constrain it to be an _NDMOBJECT

and for other things we can use super classes and constrain to those I think

**Comment by joshStillerman** [comment]
testing now.

   Looks like it works

**Comment by jcwright77** [comment]
Great. I’m a little surprised it checks the class hierarchy and not just the class name. But great that it works.

Trip is going well.  Be back Tuesday.

-john

On Mar 6, 2018, at 9:05 PM, Josh Stillerman <notifications@github.com<mailto:notifications@github.com>> wrote:

testing now.

  Looks like it works
